### PR TITLE
[MPS] Add complex->bool specialization

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #184740
* __->__ #185938

`cast_to<T>` for a complex source returned `static_cast<T>(from.x)`, using only the real part. That is correct for complex->float/int (PyTorch discards the imaginary part), but wrong for complex->bool, where the result must be a nonzero test over the whole value: `bool(0+1j)` is `True`, not `False`.

Which can be observed by running the following
```
python -c "import torch; z = torch.tensor([0+1j, 1+0j, 0+0j, 0-0.001j]);print(z.to(torch.bool), z.to(torch.float))"
```

Allows one to remove expected failure from `_refs.logical_not` for MPS complex64, which was masking this bug and now passes.

Authored with Claude.

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Comment by malfet** [comment]
> @pytorchbot rebase -b main

This will break the stack though...

**Comment by huydhn** [comment]
> This will break the stack though...

Oh, interesting, I thought rebase bot could handle ghstack
